lib/scraped_page_archive.rb

```diff
@@ -14,41 +14,100 @@
+    end
+
+    def git
+      @git ||= Git.clone(git_url, tmpdir).tap do |g|
+        g.config('user.name', "scraped_page_archive gem #{ScrapedPageArchive::VERSION}")
+        g.config('user.email', "scraped_page_archive-#{ScrapedPageArchive::VERSION}@scrapers.everypolitician.org")
+        if g.branches[branch_name] || g.branches["origin/#{branch_name}"]
+          g.checkout(branch_name)
+        else
+          g.chdir do
+            # FIXME: It's not currently possible to create an orphan branch with ruby-git
+            # @see https://github.com/schacon/ruby-git/pull/140
+            system("git checkout --orphan #{branch_name}")
+            system("git rm --quiet -rf .")
+          end
+          g.commit("Initial commit", allow_empty: true)
+        end
+      end
+    end
+
+    def tmpdir
+      @tmpdir ||= Dir.mktmpdir
+    end
+
+    def git_url
+      @git_url ||= begin
+        url = URI.parse(github_repo_url)
+        url.password = ENV['SCRAPED_PAGE_ARCHIVE_GITHUB_TOKEN']
+        url.to_s
+      end
+    end
+
+    def github_repo_url
+      @github_repo_url ||= (ENV['MORPH_SCRAPER_CACHE_GITHUB_REPO_URL'] || git_remote_get_url_origin)
+    end
+
+    def git_remote_get_url_origin
+      remote_url = `git config remote.origin.url`.chomp
+      return nil unless $?.success?
+      remote_url
+    end
+
+  end
 
   def self.record(*args, &block)
-    new.record(*args, &block)
+    new(GitStorage.new).record(*args, &block)
+  end
+
+  attr_reader :storage
+
+  def initialize(storage)
+    @storage = storage
   end
 
   def record(&block)
-    if github_repo_url.nil?
+    if storage.github_repo_url.nil?
       warn "Could not determine git repo for 'scraped_page_archive' to use.\n\n" \
         "See https://github.com/everypolitician/scraped_page_archive#usage for details."
       return block.call
     end
-    VCR::Archive::Persister.storage_location = git.dir.path
+    VCR::Archive::Persister.storage_location = storage.git.dir.path
     ret = VCR.use_cassette('', &block)
-
-    # NOTE: This is a workaround for a ruby-git bug.
-    # @see https://github.com/schacon/ruby-git/issues/23
-    git.status.changed.each { git.diff.entries }
-
-    files = (git.status.changed.keys + git.status.untracked.keys)
-    return ret unless files.any?
-    # For each interaction, commit the yml and html along with the correct commit message.
-    files.find_all { |f| f.end_with?('.yml') }.each do |f|
-      interaction = git.chdir { YAML.load_file(f) }
-      message = "#{interaction['response']['status'].values_at('code', 'message').join(' ')} #{interaction['request']['uri']}"
-      git.add([f, f.sub(/\.yml$/, '.html')])
-      git.commit(message)
-    end
-    # FIXME: Auto-pushing should be optional if the user wants to manually do it at the end.
-    git.push('origin', branch_name)
+    storage.save
     ret
   end
 
   def open_from_archive(url, *args)
-    git.chdir do
+    storage.git.chdir do
```

Comment:
Hmm, yes, that wasn't so good was it! I've added a `chdir` and `path` method to the storage class in https://github.com/everypolitician/scraped_page_archive/commit/70e492664da2c46ed851a9250f6bca2e2bb11a7a, that reduces the amount of unnecessary knowledge of the inner workings of this class. There is still a [call to `storage.github_repo_url`](https://github.com/everypolitician/scraped_page_archive/blob/70e492664da2c46ed851a9250f6bca2e2bb11a7a/lib/scraped_page_archive.rb#L28-L32), but I think that's OK to leave there for now and tidy up in a future PR as we need to give a bit more thought to moving the error message that goes with that code.